capsules/extra/src/ieee802154/driver.rs

```diff
@@ -1042,12 +1088,71 @@ impl<'a, M: device::MacDevice<'a>> device::RxClient for RadioDriver<'a, M> {
                 .get_readwrite_processbuffer(rw_allow::READ)
                 .and_then(|read| {
                     read.mut_enter(|rbuf| {
-                        let len = min(rbuf.len(), data_offset + data_len);
-                        // Copy the entire frame over to userland, preceded by two
-                        // bytes: the data offset and the data length.
-                        rbuf[..len].copy_from_slice(&buf[..len]);
-                        rbuf[0].set(data_offset as u8);
-                        rbuf[1].set(data_len as u8);
+                        ///////////////////////////////////////////////////////////////////////////////////////////
+                        // NOTE: context for the ring buffer and assumptions regarding the ring buffer
+                        // format and usage can be found in the detailed comment at the top of this file.
+                        //      Ring buffer format:
+                        //          | read index | write index | user_frame 0 | user_frame 1 | ... | user_frame n |
+                        //      user_frame format:
+                        //          | data_offset | data_len | mic_len | 15.4 frame |
+                        ///////////////////////////////////////////////////////////////////////////////////////////
+
+                        // 2 bytes for the readwrite buffer metadata (read / write index)
+                        const RING_BUF_METADATA_SIZE: usize = 2;
+
+                        // Confirm the availability of the buffer. A buffer of len 0 is indicative
+                        // of the userprocess not allocating a readwrite buffer. We must also
+                        // confirm that the userprocess correctly formatted the buffer to be of length
+                        // 2 + n * USER_FRAME_MAX_SIZE, where n is the number of user frames that the
+                        // buffer can store.
+                        if rbuf.len() == 0
+                            || (rbuf.len() - RING_BUF_METADATA_SIZE) % USER_FRAME_MAX_SIZE != 0
```

Comment:
What happens if `rbuf.len()` is 1?

Reply:
That case should fail, but it appears it may slip through since `rbuf.len()` and `RING_BUF_METADATA_SIZE` are both unsigned. I'll update this to account for that case. Good catch!